### PR TITLE
add multiple group delete functionality

### PR DESCRIPTION
**Pull request by christiemolloy**
## Motivation
https://issues.redhat.com/browse/APPDUX-556

## Brief Description
Add multi delete functionality from the toolbar

## Verification Steps
1. Go to Groups 
2. Select multiple table rows
3. Click on the top kebab and click 'delete'
4. This should delete all the table rows that were selected and refresh
5. Create a group and then try to delete multiple groups in different rows.
6. Delete a group from the table row and then try to delete multiple rows from toolbar.

## Checklist:

- [ ] Code has been tested locally by PR requester
- [ ] User-visible strings are using the react-i18next framework (useTranslation)
- [ ] Help has been implemented
- [ ] axe report has been run and resulting a11y issues have been resolved
- [ ] Unit tests have been created/updated
- [ ] Formatting has been performed via prettier/eslint
- [ ] Type checking has been performed via 'yarn check-types'

### CODE REVIEW

File: src/groups/GroupsSection.tsx [hunk 1]
```diff
@@ -31,20 +34,27 @@ export const GroupsSection = () => {
...
 
   const loader = async () => {
     const groups = await httpClient.doGet<ServerGroupsArrayRepresentation[]>(
-      "/admin/realms/master/groups"
+      `/admin/realms/${realm}/groups`
     );
     const groupsData = groups.data!;
-
+    console.log(
+      "what is groupsData here" + groupsData + JSON.stringify(groupsData)
```
Comment (edewit) [inline src/groups/GroupsSection.tsx:52]: Guess we can remove this now that it works 😄

**Review by edewit — CHANGES_REQUESTED** [review]

File: src/groups/GroupsSection.tsx [hunk 2]
```diff
@@ -90,6 +100,29 @@ export const GroupsSection = () => {
...
+    const deleteGroup = async (rowId: number) => {
+      try {
+        await httpClient.doDelete(
+          `/admin/realms/${realm}/groups/${
+            filteredData ? filteredData : rawData![rowId].id
+          }`
+        );
+        loader();
+      } catch (error) {
+        addAlert(`${t("clientDeleteError")} ${error}`, AlertVariant.danger);
```
Comment (ssilvert) [inline src/groups/GroupsSection.tsx:113]: `clientDeleteError` is missing from the resource bundle.

**Review by ssilvert — CHANGES_REQUESTED** [review]
Looks good.  Just a couple of small changes and it should be ready to go.

**Comment by christiemolloy** [comment]
updated @edewit @ssilvert

File: src/groups/messages.json [hunk 3]
```diff
@@ -17,8 +16,10 @@
     "createAGroup": "Create a group",
     "create": "Create",
     "noSearchResults": "No search results",
-    "noSearchResultsInstructions" : "Click on the search bar above to search for groups",
-    "noGroupsInThisRealm" : "No groups in this Realm",
-    "noGroupsInThisRealmInstructions" : "You haven't created any groups in this realm. Create a group to get started."
+    "noSearchResultsInstructions": "Click on the search bar above to search for groups",
+    "noGroupsInThisRealm": "No groups in this Realm",
```
Comment (srambach) [inline src/groups/messages.json:20]: should Realm be lower case?

File: src/groups/messages.json [hunk 4]
```diff
@@ -17,8 +16,10 @@
...
+    "noGroupsInThisRealmInstructions": "You haven't created any groups in this realm. Create a group to get started.",
+    "groupsDeleted": "Groups deleted",
+    "groupDeleteError": "Group unsuccessfully deleted"
```
Comment (srambach) [inline src/groups/messages.json:23]: What does that mean exactly?  🤣  So... not deleted?

Comment (ssilvert) [reply src/groups/messages.json:23]: @sguilhen Good catch.  How about "Error deleting group"

**Comment by christiemolloy** [comment]
updated @srambach @ssilvert

File: src/groups/GroupsSection.tsx [hunk 5]
```diff
@@ -60,14 +67,16 @@ export const GroupsSection = () => {
...
-
+    setFilteredData(updatedObject);
     setRawData(updatedObject);
   };
 
   useEffect(() => {
     loader();
   }, []);
 
+  console.log("what is filteredData" + filteredData);
```
Comment (mfrances17) [inline src/groups/GroupsSection.tsx:78]: remove this debugging message

**Review by mfrances17 — CHANGES_REQUESTED** [review]
Code looks good, one small comment to remove a console msg. 

Also worth investigating, when I originally tested, I was getting inconsistent deletes, sometimes it only deleted one of the couple I selected, sometimes neither. It was upon first startup of the server/admin UI. Once it started working, it worked flawlessly. I'm just suspicious of those first few occurrences. I saw this in the server log:

```
16:48:31,976 ERROR [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (default task-1) Unique index or primary key violation: "SIBLING_NAMES_INDEX_E ON PUBLIC.KEYCLOAK_GROUP(REALM_ID, PARENT_GROUP, NAME) VALUES ('master', ' ', 'adfadf', 8)"; SQL statement:
insert into KEYCLOAK_GROUP (NAME, PARENT_GROUP, REALM_ID, ID) values (?, ?, ?, ?) [23505-197]
```

File: src/groups/GroupsSection.tsx [hunk 6]
```diff
@@ -161,6 +201,7 @@ export const GroupsSection = () => {
             message={t("noGroupsInThisRealm")}
             instructions={t("noGroupsInThisRealmInstructions")}
             primaryActionText={t("createGroup")}
+            onPrimaryAction={() => handleModalToggle()}
           />
```
Comment (jenny-s51) [inline src/groups/GroupsSection.tsx:205]: When I deleted all the groups to return to the empty state, the `Create Group` button didn't show the `GroupsCreateModal`. Looks like you just need to add

```  
<GroupsCreateModal
          isCreateModalOpen={isCreateModalOpen}
          handleModalToggle={handleModalToggle}
          setIsCreateModalOpen={setIsCreateModalOpen}
          createGroupName={createGroupName}
          setCreateGroupName={setCreateGroupName}
          refresh={loader}
/>
```

under `ListEmptyState` here, and wrap them in a JSX fragment for the modal to toggle.

**Review by jenny-s51 — CHANGES_REQUESTED** [review]
Great work @christiemolloy ! Just fix the Empty State toggle -- looking good overall!

File: src/groups/GroupsSection.tsx (hunk 6 above)
Comment (christiemolloy) [reply src/groups/GroupsSection.tsx:205]: Great catch!!!

**Comment by christiemolloy** [comment]
@jenny-s51 @mfrances17 updated! @ssilvert have you seen the error message that Mark is referring to in the console of the old product ?

**Review by edewit — APPROVED** [review]
Don't know what that message is about, seems strange that for an delete there is an insert on the database. I think we can merge this when there is some backend problem, which I doubt btw, we can leave that for others to fix.

**Comment by mfrances17** [comment]
I was able to reproduce the issue.

1. Created 5 groups, named mark1 - mark5.
2. Selected mark4 and mark5, selected the Delete on mark4 group table row. The mark4 group was deleted, the mark5 group remained. Two notifications of group deleted appear.
3. Select any combination of the remaining group rows (mark1, mark2, mark3) along with mark5, and select the group delete, all other rows will delete but the mark5 group row will never delete. You need to select it by itself and explicitly delete it.

**Comment by christiemolloy** [comment]
> I was able to reproduce the issue.
> 
> 1. Created 5 groups, named mark1 - mark5.
> 2. Selected mark4 and mark5, selected the Delete on mark4 group table row. The mark4 group was deleted, the mark5 group remained. Two notifications of group deleted appear.
> 3. Select any combination of the remaining group rows (mark1, mark2, mark3) along with mark5, and select the group delete, all other rows will delete but the mark5 group row will never delete. You need to select it by itself and explicitly delete it.

This was an awesome bug to find, thank you Mark!!!! I found the problem - after deleting an individual row I needed to reset the `tableRowSelectedArray` value back to an empty array. The problem was that the array still included the last clicked indexes which was causing the problem.

**Review by mfrances17 — APPROVED** [review]
LGTM... I retested and the multi-delete issue I called out is now fixed. Good job, Christie.

**Review by ssilvert — APPROVED** [review]

**Comment by ssilvert** [comment]
> LGTM... I retested and the multi-delete issue I called out is now fixed. Good job, Christie.

Great.  Let's merge it!

**Pull request merged**